client/lib/parsing.rs

```diff
@@ -666,4 +771,109 @@ mod tests {
         valid_simple_args_test(&format!("x:opt_public_key='{}'", hex_value), Some(value));
         valid_simple_args_test::<Option<PublicKey>>("x:opt_public_key=null", None);
     }
+
+    #[test]
+    fn should_fail_to_parse_bad_args() {
+        invalid_simple_args_test(bad::ARG_BAD_TYPE);
+        invalid_simple_args_test(bad::ARG_GIBBERISH);
+        invalid_simple_args_test(bad::ARG_UNQUOTED);
+        invalid_simple_args_test(bad::EMPTY);
+        invalid_simple_args_test(bad::LARGE_2K_INPUT);
+    }
+
+    #[test]
+    fn should_fail_to_parse_conflicting_arg_types() {
+        assert_eq!(
+            parse_session_info(
+                "",
+                "",
+                "",
+                "",
+                "",
+                &["something:u32='0'"],
+                "path_to/file",
+                "",
+                "",
+            )
+            .map(|_| ())
+            .map_err(ErrWrapper),
+            Err(Error::ConflictingArguments {
+                context: "parse_session_info",
+                args: vec!["session_args".to_owned(), "session_args_complex".to_owned()]
+            }
+            .into())
+        );
+        assert_eq!(
+            parse_payment_info(
+                "",
+                "",
+                "",
+                "",
+                "",
+                "",
+                &["something:u32='0'"],
+                "path_to/file",
+                "",
+                "",
+            )
+            .map(|_| ())
+            .map_err(ErrWrapper),
+            Err(Error::ConflictingArguments {
+                context: "parse_payment_info",
+                args: vec!["payment_args".to_owned(), "payment_args_complex".to_owned()]
+            }
+            .into())
+        );
+    }
+
+    #[test]
+    fn should_fail_to_parse_conflicting_session_parameters() {
```

Comment:
It'll be tedious, but I'd like to see a batch of tests corresponding to each `SessionStrParams::with_xxx` constructor.

For each, we need to check that providing mismatched args causes failure.  Taking the case again of `SessionStrParams::with_package_hash`, we need a check for each of the following:
* a non-empty `session_hash` causes failure
* a non-empty `session_name` causes failure
* a non-empty `session_package_name` causes failure
* a non-empty `session_path` causes failure
* non-empty `session_args` and non-empty `session_args_complex` causes failure
* a non-empty `session_entry_point` causes failure

Reply:
I created the macros lower in this file to address the tedium a repetition, as well as implemented the variants of those tests using them.